common/txmgr/finalizer.go

```diff
@@ -0,0 +1,235 @@
+	return f.processHead(ctx, head)
+}
+
+// Determines if any confirmed transactions can be marked as finalized by comparing their receipts against the latest finalized block
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) processHead(ctx context.Context, head types.Head[BLOCK_HASH]) error {
+	latestFinalizedHead := head.LatestFinalizedHead()
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("failed to find latest finalized head in chain")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	confirmedTxs, err := f.txStore.FindTransactionsByState(ctx, TxConfirmed, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve confirmed transactions: %w", err)
+	}
+
+	var finalizedTxs []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	// Group by block hash transactions whose receipts cannot be validated using the cached heads
+	receiptBlockHashToTx := make(map[BLOCK_HASH][]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE])
+	// Find transactions with receipt block nums older than the latest finalized block num and block hashes still in chain
+	for _, tx := range confirmedTxs {
+		// Only consider transactions not already marked as finalized
+		if tx.Finalized {
+			continue
+		}
+		receipt := tx.GetReceipt()
+		if receipt == nil || receipt.IsZero() || receipt.IsUnmined() {
+			continue
+		}
+		// Receipt newer than latest finalized head block num
+		if receipt.GetBlockNumber().Cmp(big.NewInt(latestFinalizedHead.BlockNumber())) > 0 {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.GetBlockNumber().Int64() < earliestBlockNumInChain {
+			receiptBlockHashToTx[receipt.GetBlockHash()] = append(receiptBlockHashToTx[receipt.GetBlockHash()], tx)
+			continue
+		}
+		blockHashInChain := latestFinalizedHead.HashAtHeight(receipt.GetBlockNumber().Int64())
+		// Receipt block hash does not match the block hash in chain. Transaction has been re-org'd out but DB state has not been updated yet
+		if blockHashInChain.String() != receipt.GetBlockHash().String() {
+			continue
+		}
+		finalizedTxs = append(finalizedTxs, tx)
+	}
+
+	// Check if block hashes exist for receipts on-chain older than the earliest cached head
+	// Transactions are grouped by their receipt block hash to minimize the number of RPC calls in case transactions were confirmed in the same block
+	// This check is only expected to be used in rare cases if there was an issue with the HeadTracker or if the node was down for significant time
```

Comment:
That is not true. This is directly affected by the `HistoryDepth` config we have in Head Tracker, which controls how many blocks we store before the latest finalized. This is actually a very useful observation. Initially, we were aiming to have `HistoryDepth` set to 0 since most of HT's users were not interested in blocks past the latest finalized. However, based on your approach above it makes sense to set a reasonable value for that config so we can avoid excessive RPC calls in exchange for some in-memory storage. Once we introduce HT's optimizations, we could further increase it.
CAUTION: Recently we observed on certain chains that the depth of `latest` to 'finalized` is quite large, so we added a config to  return an error when this happens. We should consider such cases here and perhaps gracefully handle them or even alert the user as transactions might not get finalized for a long period of time.

Reply:
Good to know! Didn't know we were considering setting `HistoryDepth=0`. Surprised none of the other callers needed to do a block hash check to protect against re-orgs but I don't really know the use cases. 

For the case that the depth of latest to 'finalized` becomes quite large, I can check with CCIP/Automation to see if they care. I could foresee warning on this adding special scenarios that become hard to debug especially if products act on them.